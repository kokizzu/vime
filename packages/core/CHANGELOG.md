# Change Log

All notable changes to this project will be documented in this file.

## [1.0.2](https://github.com/vime-js/vime/compare/@vime/core@1.0.1...@vime/core@1.0.2) (2020-08-13)


### Bug Fixes

* **core:** point unpkg and jsdelivr to correct file ([8360ad6](https://github.com/vime-js/vime/commit/8360ad62197ab35a3f51986f83e9412c7fd11a3b))

## [1.0.1](https://github.com/vime-js/vime/compare/@vime/core@1.0.0...@vime/core@1.0.1) (2020-08-13)


### Bug Fixes

* **core:** icons not loading ([2b48c85](https://github.com/vime-js/vime/commit/2b48c8547b0b7454b48a2c2707e82d74837ae9cb))

# 1.0.0 (2020-08-12)


### Bug Fixes

* **core:** lazy loading broken ([792520a](https://github.com/vime-js/vime/commit/792520ac17dcd8d70179ac4b29d91d4ac5a127eb))
* **core:** prevent posters loading twice with custom UI ([5aefb06](https://github.com/vime-js/vime/commit/5aefb06f3ad7e80204ce4966f42d5372a3517a70))
* **core/ui:** slider keydown events propagating ([94fafe1](https://github.com/vime-js/vime/commit/94fafe1f485e841febea3dd8a065588bf874622a))
* **core/ui/captions:** emitting events after disconnecting ([5f0abac](https://github.com/vime-js/vime/commit/5f0abac07f6b497607968265435e76a248ba3edf))
* **core/ui/captions:** toggling captions on/off broken ([090d776](https://github.com/vime-js/vime/commit/090d7762401750d91f9a1cc12a320cbfa3652eb6))


### Features

* **core:** add lazy loading via lozad ([3bde06c](https://github.com/vime-js/vime/commit/3bde06c5928895195f8e934c6808eca964737d07))
* **core:** add skeleton loading animation ([ff7860d](https://github.com/vime-js/vime/commit/ff7860ddfe74f029c3d365870a31d6d3d72f2748))
* **core:** dailymotion + hls + dash providers ([c520c30](https://github.com/vime-js/vime/commit/c520c302e5db91f20c9944915a43102ff1305e51))
* **core:** improve test harness + vimeo provider ([dd153ca](https://github.com/vime-js/vime/commit/dd153ca96026a3ef80c89fa1dcc560e36dcc7c1d))
* **core:** vime-embed component ([b45be6f](https://github.com/vime-js/vime/commit/b45be6f642e286b7da568b4450e8268334961019))
* **core:** vime-player component ([c86bee4](https://github.com/vime-js/vime/commit/c86bee4f709fd51d6457805df6c6c19efa0abc89))
* **core:** youtube provider + providers test harness ([8f93a68](https://github.com/vime-js/vime/commit/8f93a68edd16d4396c708e84e8e5bc1d8da69c41))
* **core/ui:** add click-to-play component ([fcce4f1](https://github.com/vime-js/vime/commit/fcce4f11b32f4603f96677dd10803233c344e8d8))
* **core/ui:** add light audio player theme ([93d4206](https://github.com/vime-js/vime/commit/93d42062ee2515ac37a10244961767714cfa87f9))
* **core/ui:** captions component ([c810f42](https://github.com/vime-js/vime/commit/c810f42d00cc81aa3cf35bf81baf2846dcbba4bb))
* **core/ui:** controls components ([5dc4e26](https://github.com/vime-js/vime/commit/5dc4e261e23df6993694abfa776af8637fb7b9a6))
* **core/ui:** create settings components ([3a1fd82](https://github.com/vime-js/vime/commit/3a1fd82a232b67f27759dc94e1e76fea8228f97e))
* **core/ui:** icons + icon component ([d13df8e](https://github.com/vime-js/vime/commit/d13df8ef1b265ec995b8a37f023162f7a978da37))
* **core/ui:** poster + spinner components ([cbdd69f](https://github.com/vime-js/vime/commit/cbdd69ff9f3474639f2a781809859c3b94beb7ff))
* **core/ui:** scrim component ([df47efe](https://github.com/vime-js/vime/commit/df47efe31b3088157896120b9f7ff2d78bdd5c99))
* **core/ui/controls:** add option to scale control size up/down ([3635b1a](https://github.com/vime-js/vime/commit/3635b1ae0d2d2d0163086a01d4e1f6fb2468a561))
* **core/ui/poster:** add poster events ([9d36f47](https://github.com/vime-js/vime/commit/9d36f47a55f6459e8ffc24018cf2fb5c2654fcde))
* **core/ui/scrim:** add events ([1180240](https://github.com/vime-js/vime/commit/1180240c8e93cce96009c3c2098064806dfa4c37))
* **core/ui/scrim:** add gradient option ([adf6bcb](https://github.com/vime-js/vime/commit/adf6bcbdb2d8520552bb6c93b5bc1f30ab95ce84))
* **core/ui/settings:** add audio/mobile styles to settings ([c937312](https://github.com/vime-js/vime/commit/c937312b344e2a5b3bb10315367f98c6b0eb81a0))
* **core/ui/spinner:** add events ([13dfd14](https://github.com/vime-js/vime/commit/13dfd1445d72357a1ed315cd3a25e955be19a772))
