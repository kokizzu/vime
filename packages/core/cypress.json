{
  "baseUrl": "http://localhost:3333/src/components",
  "integrationFolder": "cypress/tests",
  "testFiles": "**/*.e2e.ts",
  "fixturesFolder": false,
  "supportsFolder": "cypress/support/index.ts",
  "ignoreTestFiles": [
    "**/__snapshots__/*",
    "**/__image_snapshots__/*"
  ]
}
