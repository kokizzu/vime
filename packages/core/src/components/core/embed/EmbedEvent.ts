export enum EmbedEvent {
  SrcChange = 'embedSrcChange',
  Message = 'embedMessage',
  Loaded = 'embedLoaded'
}

export interface EmbedEventPayload {
  [EmbedEvent.SrcChange]: string,
  [EmbedEvent.Loaded]: void,
  [EmbedEvent.Message]: any
}
