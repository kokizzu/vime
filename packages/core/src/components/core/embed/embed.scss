vime-embed > iframe {
  position: absolute;
  top: 0;
  left: 0;
  border: 0;
  width: 100%;
  height: 100%;
  user-select: none;
  z-index: $z-index-media;
}