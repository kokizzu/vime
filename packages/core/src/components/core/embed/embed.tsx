import {
  Component,
  ComponentInterface,
  Element,
  Event,
  EventEmitter,
  h,
  Listen,
  Method,
  Prop,
  State,
  Watch,
} from '@stencil/core';
import { isString } from '../../../utils/unit';
import { appendParamsToURL, Params, preconnect } from '../../../utils/network';

let idCount = 0;
const connected = new Set();

@Component({
  tag: 'vime-embed',
  styleUrl: 'embed.scss',
})
export class Embed implements ComponentInterface {
  private id!: string;

  private iframe?: HTMLIFrameElement;

  @Element() el!: HTMLVimeEmbedElement;

  @State() srcWithParams = '';

  @State() hasEnteredViewport = false;

  /**
   * A URL that will load the external player and media (Eg: https://www.youtube.com/embed/DyTCOwB0DVw).
   */
  @Prop() embedSrc = '';

  /**
   * The title of the current media so it can be set on the inner `iframe` for screen readers.
   */
  @Prop() mediaTitle = '';

  /**
   * The parameters to pass to the embedded player which are appended to the `embedSrc` prop. These
   * can be passed in as a query string or object.
   */
  @Prop({ attribute: 'params' }) params: string | Params = '';

  @Watch('embedSrc')
  @Watch('params')
  srcChange() {
    this.srcWithParams = appendParamsToURL(this.embedSrc, this.params);
  }

  @Watch('srcWithParams')
  srcWithParamsChange() {
    if (!this.hasEnteredViewport && !connected.has(this.embedSrc)) {
      if (preconnect(this.srcWithParams)) connected.add(this.embedSrc);
    }

    this.vEmbedSrcChange.emit(this.srcWithParams);
  }

  /**
   * Where the src request had originated from without any path information.
   */
  @Prop() origin?: string;

  /**
   * A collection of URLs to that the browser should immediately start establishing a connection
   * with.
   */
  @Prop({ attribute: 'preconnections' }) preconnections: string[] = [];

  /**
   * A function which accepts the raw message received from the embedded media player via
   * `postMessage` and converts it into a POJO.
   */
  @Prop({ attribute: 'decoder' }) decoder?: (data: string) => Params | undefined;

  /**
   * Emitted when the `embedSrc` or `params` props change. The payload contains the `params`
   * serialized into a query string and appended to `embedSrc`.
   */
  @Event({ bubbles: false }) vEmbedSrcChange!: EventEmitter<string>;

  /**
   * Emitted when a new message is received from the embedded player via `postMessage`.
   */
  @Event({ bubbles: false }) vEmbedMessage!: EventEmitter<any>;

  /**
   * Emitted when the embedded player and any new media has loaded.
   */
  @Event({ bubbles: false }) vEmbedLoaded!: EventEmitter<void>;

  @Watch('preconnections')
  preconnectionsChange() {
    if (this.hasEnteredViewport) { return; }

    this.preconnections
      .filter((connection) => !connected.has(connection))
      .forEach((connection) => {
        if (preconnect(connection)) connected.add(connection);
      });
  }

  connectedCallback() {
    this.srcChange();
    this.genIframeId();
  }

  @Listen('message', { target: 'window' })
  onWindowMessage(e: MessageEvent) {
    const originMatches = (e.source === this.iframe?.contentWindow)
      && (!isString(this.origin) || this.origin === e.origin);

    if (!originMatches) return;

    const message = this.decoder?.(e.data) ?? e.data;
    if (message) this.vEmbedMessage.emit(message);
  }

  /**
   * Posts a message to the embedded media player.
   */
  @Method()
  async postMessage(message: any, target?: string) {
    this.iframe?.contentWindow?.postMessage(JSON.stringify(message), (target ?? '*'));
  }

  private onLoad() {
    this.vEmbedLoaded.emit();
  }

  private genIframeId() {
    idCount += 1;
    this.id = `vime-iframe-${idCount}`;
  }

  render() {
    return (
      <iframe
        class="lazy"
        id={this.id}
        title={this.mediaTitle}
        data-src={this.srcWithParams}
        // @ts-ignore
        allowfullscreen="1"
        allow="autoplay; encrypted-media; picture-in-picture"
        onLoad={this.onLoad.bind(this)}
        ref={(el: any) => { this.iframe = el; }}
      />
    );
  }
}
