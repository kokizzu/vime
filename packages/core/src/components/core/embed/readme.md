# vime-embed

Embeds an external media player and enables interacting with it via `postMessage`.

## Example

```html
<vime-embed
  embed-src="https://www.youtube-nocookie.com/embed/DyTCOwB0DVw"
  media-title="Agent 327: Operation Barbershop"
  origin="https://www.youtube-nocookie.com"
/>

<script>
  const embed = document.querySelector('vime-embed');
  embed.params = { autoplay: 1, muted: 1, controls: 0 };
  embed.addEventListener('embedMessage', (e) => {
    const message = e.detail;
  });
</script>
```

<!-- Auto Generated Below -->


## Properties

| Property         | Attribute     | Description                                                                                                                     | Type                                                                | Default     |
| ---------------- | ------------- | ------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------- | ----------- |
| `decoder`        | --            | A function which accepts the raw message received from the embedded media player via `postMessage` and converts it into a POJO. | `((data: string) => Record<string, any> \| undefined) \| undefined` | `undefined` |
| `embedSrc`       | `embed-src`   | A URL that will load the external player and media (Eg: https://www.youtube.com/embed/DyTCOwB0DVw).                             | `string`                                                            | `''`        |
| `mediaTitle`     | `media-title` | The title of the current media so it can be set on the inner `iframe` for screen readers.                                       | `string`                                                            | `''`        |
| `origin`         | `origin`      | Where the src request had originated from without any path information.                                                         | `string \| undefined`                                               | `undefined` |
| `params`         | --            | The parameters to pass to the embedded player. These are encoded as a query string and appended to the `embedSrc` prop.         | `{ [x: string]: any; }`                                             | `{}`        |
| `preconnections` | --            | A collection of URLs to that the browser should immediately start establishing a connection with.                               | `string[]`                                                          | `[]`        |


## Events

| Event            | Description                                                                                                                                        | Type                  |
| ---------------- | -------------------------------------------------------------------------------------------------------------------------------------------------- | --------------------- |
| `embedLoaded`    | Emitted when the embedded player and any new media has loaded.                                                                                     | `CustomEvent<void>`   |
| `embedMessage`   | Emitted when a new message is received from the embedded player via `postMessage`.                                                                 | `CustomEvent<any>`    |
| `embedSrcChange` | Emitted when the `embedSrc` or `params` props change. The payload contains the `params` serialized into a query string and appended to `embedSrc`. | `CustomEvent<string>` |


## Methods

### `postMessage(message: any, target?: string | undefined) => Promise<void>`

Posts a message to the embedded media player.

#### Returns

Type: `Promise<void>`




## Dependencies

### Used by

 - [vime-dailymotion](../../providers/dailymotion)
 - [vime-vimeo](../../providers/vimeo)
 - [vime-youtube](../../providers/youtube)

### Graph
```mermaid
graph TD;
  vime-dailymotion --> vime-embed
  vime-vimeo --> vime-embed
  vime-youtube --> vime-embed
  style vime-embed fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
