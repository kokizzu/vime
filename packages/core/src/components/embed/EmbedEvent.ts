import { Params } from '../../utils/network';

export enum EmbedEvent {
  SrcChange = 'vimeEmbedSrcChange',
  Message = 'vimeEmbedMessage',
  Loaded = 'vimeEmbedLoaded'
}

export interface EmbedEventPayload {
  [EmbedEvent.SrcChange]: string,
  [EmbedEvent.Loaded]: void,
  [EmbedEvent.Message]: Params
}
