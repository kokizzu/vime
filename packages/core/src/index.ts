import { PlayerProp, PlayerProps } from './components/core/player/PlayerProp';
import { PlayerDispatcher, createPlayerDispatcher } from './components/core/player/PlayerDispatcher';
import { withPlayerContext, usePlayerContext } from './components/core/player/PlayerContext';
import { PlayerEvent } from './components/core/player/PlayerEvent';
import { findRootPlayer } from './components/core/player/utils';

export * from './components';

export {
  PlayerProp,
  PlayerProps,
  PlayerEvent,
  findRootPlayer,
  withPlayerContext,
  usePlayerContext,
  PlayerDispatcher,
  createPlayerDispatcher,
};
