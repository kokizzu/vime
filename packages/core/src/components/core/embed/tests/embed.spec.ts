import { newSpecPage, SpecPage } from '@stencil/core/testing';
import { Embed } from '../embed';
import { EmbedEvent } from '../EmbedEvent';

let page: SpecPage;

beforeEach(async () => {
  page = await newSpecPage({
    components: [Embed],
    html: '<vime-embed></vime-embed>',
  });
});

it('should be structurally sound', () => {
  expect(page.root).toMatchSnapshot();
});

it('should increment iframe id', async () => {
  page = await newSpecPage({
    components: [Embed],
    html: '<vime-embed></vime-embed><vime-embed></vime-embed>',
  });
  const frames = page.doc.querySelectorAll('iframe');
  expect(frames).toHaveLength(2);
  // Starts at 3 because of the 2 prior tests.
  expect(frames[0].id).toEqual('vime-iframe-3');
  expect(frames[1].id).toEqual('vime-iframe-4');
});

it('should fire event when src/params change', async () => {
  const cb = jest.fn();
  page.root!.addEventListener(EmbedEvent.SrcChange, cb);
  page.rootInstance!.embedSrc = 'http://apples.com';
  await page.waitForChanges();
  expect(cb).toHaveBeenCalled();
  page.rootInstance!.params = { controls: 0 };
  await page.waitForChanges();
  expect(cb).toHaveBeenCalledTimes(2);
});

it('should set the iframe title', async () => {
  const iframe = page.root!.querySelector('iframe');
  expect(iframe?.title).toEqual('');
  page.root!.mediaTitle = 'Apples!';
  await page.waitForChanges();
  expect(iframe?.title).toEqual('Apples!');
});

it('should preconnect', async () => {
  page.rootInstance!.preconnections = [
    'https://apples.com',
    'https://bees.com',
  ];
  await page.waitForChanges();
  const links = page!.doc.head.querySelectorAll('link');
  expect(links[0]?.href).toEqual('https://apples.com/');
  expect(links[1]?.href).toEqual('https://bees.com/');
});
