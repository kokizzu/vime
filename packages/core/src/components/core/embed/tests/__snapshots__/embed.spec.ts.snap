// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be structurally sound 1`] = `
<vime-embed>
  <iframe allow="autoplay; encrypted-media; picture-in-picture" allowfullscreen="1" id="vime-embed-1" src=""></iframe>
</vime-embed>
`;
